Network Skills Okep (#52)
ovn-kubernetes/ovn-kubernetes-mcp: docs/okeps/okep-59-ovn-kubernetes-network-skills.md

  same PR; CI rejects PRs that leave a skill referencing a removed tool.

# Implementation Details

Phase-1 deliverables:

1. Move the drafts under
   [`hack/network_skills/.cursor/skills/`](../hack/network_skills/.cursor/skills/)
   to canonical `skills/`. Add `skills/SKILLS.md` (index + format
   reference).
2. Add `.cursor/skills` and `.claude/skills` symlinks.
3. Add `make skills-lint`: parses each `SKILL.md`, validates frontmatter,
   checks every tool reference against the in-process MCP tool registry
   (`RegisteredToolNames()` already exists per OKEP-5494), rejects
   free-form bash and write verbs, enforces the reporting block and the
   line-count cap.
4. Add a short docs page on pointing Cursor / Claude Code / `@cursor/sdk`
   at the server and the available skills.

**Tool registry coupling.** The MCP server and the skills lint share the
same `RegisteredToolNames()` source of truth. Renaming a tool is a CI
failure on every skill that references it - not a runtime regression.

**Skills as MCP resources (Phase 2).** When ready, add a resource provider
exposing each skill at e.g. `mcp://ovn-kubernetes/skills/<name>` with

> `mimeType: text/markdown`. Additive; does not change the file format.

Author: verify

# OKEP-59: Network Troubleshooting Skills for OVN-Kubernetes

* Issue: [#59](https://github.com/ovn-kubernetes/ovn-kubernetes-mcp/issues/59)

> Companion to [OKEP-5494: Model Context Protocol for Troubleshooting OVN-Kubernetes](./okep-5494-ovn-kubernetes-mcp-server.md).
> OKEP-5494 gives the LLM safe, layered *tools*. This OKEP gives it expert

> > *guidance* on how to use them.

Reviewer: If you wanto refer to the original OKEP use the web link: https://ovn-kubernetes.io/okeps/okep-5494-ovn-kubernetes-mcp-server/

Reviewer: Also I am not sure why the original OKEP has to be mentioned. The original OKEP is already implemented and this OKEP enhances on the current state of the MCP server. Ideally, this OKEP should mention what is missing in the current state of the MCP server rather than referring the original OKEP.

Author: done, my bad as replied i misunderstood one of the above comments

  network triage shipped alongside the MCP server.
* Adopt the cross-vendor [Agent Skills](https://www.anthropic.com/news/skills)
  convention (`SKILL.md` with `name` + `description` frontmatter) so skills
  work in any compatible runtime (Cursor, Cursor CLI, Claude Code,
  Anthropic Skills API, custom `@cursor/sdk` agents) without forking
  per-client.
* Ship an **initial set of few skills** covering the most frequent
  Phase-1 triage paths: `pod-to-pod-connectivity`,
  `service-connectivity`, `network-policy-debugging`, `dns-resolution`,
  `external-connectivity`, `node-networking`, `ovn-topology-discovery`,
  `ovs-flow-analysis`, `conntrack-debugging`, `packet-drop-investigation`.
* Skills MUST only call **named MCP tools** from OKEP-5494 (no free-form
  bash), MUST be **read-only**, and MUST encode layered, evidence-driven
  workflows.
* Skills MUST live in-tree, be code-reviewed, and be CI-linted against the
  MCP tool registry so they cannot drift.

# Future Goals

* **Feature-specific skills**: ovn-k8s skills related to features in the main product `egressip`,
  `egress-firewall`, `route-advertisements`, `ipsec`, `bgp`,

>   `udn-primary-network`, `staleSNAT`, `multus-secondary-networks`, etc.

Reviewer: Not sure if we claim `staleSNAT` as a feature of ovn-kubernetes 😃

Author: ack, removed stale SNAT is not upstream feature, thanks for correcting it.

* **Not a tool replacement.** Skills are guidance; tools live in
  OKEP-5494. Without the server, a skill is just a runbook.
* **No write / remediation.** Phase 1 stays read-only.
* **Not solving LLM quality.** A bad model + a good skill still fails;
  skills bound the failure mode, they do not eliminate it.

# Introduction

## Why skills, not just tool descriptions or a system prompt?

Tool descriptions cannot encode:

1. **Methodology.** "Always check the pod is Running before running
   `ovn-trace`" is a workflow rule, not a property of any single tool.
2. **Cross-tool wiring.** `addresses` from `ovn-get Logical_Switch_Port`
   becomes the `eth.src`/`ip4.src` of `ovn-trace`. That graph lives
   between tools, not inside any one of them.
3. **Decision trees.** "If `ovn-trace` allows but `ofproto-trace`
   drops, suspect flow-installation; restart `ovn-controller`" is
   exactly what skills are for.
4. **Context economics.** A monolithic system prompt covering all 10
   scenarios is in context for every conversation. Agent Skills load
   only the matching skill body via *progressive disclosure*: the
   `description` is always loaded; the body is loaded only when the

>    user's symptom matches.

Reviewer: Can you link a source for this behavior? Is this same across all the agents?

Author: Source is the [Anthropic Agent Skills](https://www.anthropic.com/news/skills) convention. Same lazy-load behaviour is implemented today by Claude Code, Cursor / Cursor CLI, and the Anthropic API skills feature. they all keep the SKILL.md frontmatter (name + description) in context and load the body only when the agent's routing selects that skill.

Not all agents do the same, but most of the popular ones have already agreed with this convention

   scenarios is in context for every conversation. Agent Skills load
   only the matching skill body via *progressive disclosure*: the
   `description` is always loaded; the body is loaded only when the

>    user's symptom matches.

Author: replaced it with N, the number could change with time

              | references --------->  tcpdump, pwru, ... |
              | tools by name   |   |                     |
              +-----------------+   +---------------------+
                       \                  /
                        v                v
                       ovnkube-node pods, NB/SB DBs,
                       OVS, kernel, NIC
```

A skill is a deterministic recipe over the MCP tool catalogue. If the
catalogue changes, the affected skills change in the same PR -
enforced by CI.

## Initial skill catalogue thoughts

| Skill | When to use | Primary MCP tools |
|---|---|---|
| `pod-to-pod-connectivity`   | Pod A can't reach Pod B | resource-get, ovn-show, ovn-get LSP/PB/Chassis, ovn-trace, ovs-appctl-ofproto-trace, get-conntrack, tcpdump, pwru |
| `service-connectivity`      | ClusterIP/NodePort/LB unreachable, missing endpoints | resource-get Service/EndpointSlice, ovn-get Load_Balancer, ovn-trace, ovs-appctl-dump-conntrack, get-iptables, get-nft |
| `network-policy-debugging`  | Traffic unexpectedly blocked/allowed | resource-list NetworkPolicy, ovn-get ACL/Address_Set/Port_Group, ovn-trace, ovn-lflow-list, ovs-appctl-ofproto-trace |

> | `dns-resolution`            | NXDOMAIN, DNS timeout, kube-dns unreachable | resource-get Service/Pod/ConfigMap, pod-logs CoreDNS, tcpdump :53, ovn-get Load_Balancer, get-conntrack udp/53 |

Reviewer: Is this a core ovn-kubernetes problem?

Author: No, in short, its DNS (CoreDNS) problem that only touches ovn-k through the Service load-balancer path (which is already covered by service-connectivity). Removed dns-resolution for now

| `node-networking`           | NetworkUnavailable, tunnel down, missing routes | resource-get Node, get-ip address/link/route/rule/neigh, ovs-vsctl-show, ovn-get Chassis, get-iptables, get-nft, pod-logs ovn-controller |
| `ovn-topology-discovery`    | Map / understand the logical network | ovn-show NB/SB, ovn-get LS/LR/LSP/LRP/Chassis/Port_Binding/NAT/Static_Route/Load_Balancer, ovs-vsctl-show |
| `ovs-flow-analysis`         | Logical vs OpenFlow mismatch, stale flows | ovn-lflow-list, ovs-ofctl-dump-flows, ovs-appctl-ofproto-trace, ovn-trace, pod-logs ovn-controller |
| `conntrack-debugging`       | NAT broken, stale entries, table-full, zone conflicts | ovs-appctl-dump-conntrack, get-conntrack -L/-S/-C, ovs-ofctl-dump-flows ct, ovn-lflow-list ct |
| `packet-drop-investigation` | Silent drops, unknown drop point | tcpdump, ovs-ofctl-dump-flows drop, ovs-appctl-ofproto-trace, ovn-trace, ovs-appctl-dump-conntrack, get-iptables, get-nft, pwru |

Skills are independent (pick one by symptom) and **composable** -
`dns-resolution` calls into `service-connectivity`,
`external-connectivity` into `node-networking`,
`packet-drop-investigation` into `ovs-flow-analysis` and

> `conntrack-debugging`.

Reviewer: This sentence seems a little bit contradictory. If `external-connectivity` uses `node-networking`, isn't `external-connectivity` dependent on `node-networking`?

Author: "independent" means each skill is a valid entry point from a user symptom (no ordering assumption between skills, you don't have to run node-networking before external-connectivity). "Composable" means a skill's decision tree may hand off mid-triage to a sibling when the failure crosses their boundary. So external-connectivity doesn't structurally depend on node-networking; it just delegates to it at runtime if the north-south trace surfaces a node-level problem. 
Happy to reword if the current phrasing still reads contradictory.

Reviewer: So, you mean that `external-connectivity` may load sibling skill `node-networking` if mid-triage it finds the symptoms matching of `node-networking`, but may load some other skill if the symptom matches the other one? If that is the case, this seems a very deterministic statement. It would be better to use `may` here and give a little bit more explanation.

Author: done reworded and added some explanation

# User Stories

**As an OVN-Kubernetes developer**, I want the AI assistant to follow the
layered triage methodology I would (pod state → NB → SB → OVS → kernel →
capture) so I get a consistent, auditable RCA instead of a model
improvising from training data.

**As a new engineer joining the team**, I want a published skill I can
read or run so I learn the canonical debug flow without shoulder-surfing
a senior engineer.

**As a triage engineer**, I want the agent to fetch parameters from earlier
tool output (MACs, port names, datapath names) instead of asking me to
type them.

**As a maintainer reviewing a bug**, I want the LLM transcript to map 1:1
to a published skill so I can audit each step and improve the skill in the

> same PR that fixes the bug.

Reviewer: I am not sure if this is a valid user story. The bug would be related to ovn-kubernetes and the fix for the bug will reside in the ovnk repo and not the ovnk-mcp repo. So how will the skill be fixed in the same PR?

Author: removed this user story, I feel this might be too ambitious as of now

**As a layered-feature owner** (EgressIP, IPSec, BGP, RouteAdvertisements,
UDN, ...), I want to publish a skill for my feature's failure modes so the
LLM does not have to rediscover them.

# Proposed Solution

## Repository layout

```text
ovn-kubernetes-mcp/
  skills/                              <- canonical, vendor-neutral
    SKILLS.md                          <- index + format reference
    pod-to-pod-connectivity/SKILL.md
    service-connectivity/SKILL.md
    ...
  .cursor/skills -> ../skills          <- symlink for Cursor / Cursor CLI

>   .claude/skills -> ../skills          <- symlink for Claude Code

Reviewer: Why are the symlinks needed? Can't the parent directory be used directly?

Author: This is a good question, I'll break down this in three scenarios
1 - Cursor

User does | With our symlinks | Without symlinks
-- | -- | --
Clones repo, opens in Cursor | Just works (project-scoped .cursor/skills → skills) | User adds <repo>/skills to Cursor config (30-sec, one-time)

2 - Claude Code

Where it looks by default:

~/.claude/skills/ — user-global
<project>/.claude/skills/ — project-scoped
How to add a non-default path:

.claude/settings.json at repo root: {"skillsPaths": ["../skills", ...]}
Or user-level: ~/.claude/settings.json
Loading model: identical to Cursor — frontmatter in system prompt, body lazy-loaded.

Our repo, three ways: essentially the same table as Cursor, just swap .cursor/skills → .claude/skills and .cursor/config.json → .claude/settings.json. Same trade-off.

3 - custom agent 
Any agent built on @anthropic-ai/sdk, openai, @cursor/sdk, or a home-grown runtime. 
Doesn't care about .cursor/skills/ or .claude/skills/ at all. They just need a stable canonical path — skills/ fits perfectly. Symlinks are noise from a custom-agent perspective.

The symlinks buy exactly one thing: zero-config auto-discovery for Cursor & Claude Code users on Linux/macOS who clone the repo. That's a genuine convenience, but:

It's a one-time 30-second config for the same users to add <repo>/skills to their client
It doesn't help custom agents at all
It doesn't help the binary-only install case (which is where the real zero-config story lives, Phase-2 MCP resources)

Reviewer: If the benefits are not that much, then I think it's better to not include it in the OKEP. If in future, this is needed to be done, then we can take a call at that time.

  same PR; CI rejects PRs that leave a skill referencing a removed tool.

# Implementation Details

Phase-1 deliverables:

1. Author the initial set of `SKILL.md` files under canonical `skills/`
   (one directory per skill from the catalogue above). Add
   `skills/SKILLS.md` as the index + format reference.
2. Add `.cursor/skills` and `.claude/skills` symlinks.
3. Add `make skills-lint`: parses each `SKILL.md`, validates frontmatter,
   checks every tool reference against the in-process MCP tool registry
   (`RegisteredToolNames()` already exists per OKEP-5494), rejects
   free-form bash and write verbs, enforces the reporting block and the
   line-count cap.
4. Add a short docs page on pointing Cursor / Claude Code / `@cursor/sdk`
   at the server and the available skills.

**Tool registry coupling.** The MCP server and the skills lint share the
same `RegisteredToolNames()` source of truth. Renaming a tool is a CI
failure on every skill that references it - not a runtime regression.

**Skills as MCP resources (Phase 2).** When ready, add a resource provider
exposing each skill at e.g. `mcp://ovn-kubernetes/skills/<name>` with

> `mimeType: text/markdown`. Additive; does not change the file format.

Reviewer: Will the loading of the skills will still be done like phase-1? That is, only the description will be loaded initially and the body will be loaded when the prompt matches the description? Or will the whole skill will be loaded upfront?

Author: Same as Phase-1. only the description is loaded upfront, the body is fetched on-match. MCP's resources primitive splits this naturally: resources/list returns each skill's uri + name + description (that's what the agent keeps in context for routing), and resources/read <uri> returns the body (called only when the agent picks that skill). So the whole skill is not loaded upfront.

# Security Model

Skills inherit OKEP-5494's model unchanged. They neither expand nor relax
it:

* Skills only call **named MCP tools**. They cannot escape the tool
  sandbox.
* Skills are **read-only by construction** - lint rejects write verbs.
* Skills are **public**, reviewed in the open. No secrets, no IPs,
  no customer data; placeholders only (`<src-pod>`, `<node>`, `<ovn-pod>`).
* Skills carry no RBAC of their own. They run with whatever ServiceAccount
  / kubeconfig the MCP server uses.
* **Remediation** appears only as text in the Reporting block, addressed
  to the human. The agent does not execute it under this OKEP.

| Threat | Mitigation |
|---|---|
| Malicious skill PR adds bash/exec | CI lint forbids free-form shell; PR review |

> | Stale skill references removed tool | CI lint cross-checks the live tool registry |

Reviewer: How will the lint take care of these threats? Can you elaborate on this? Is there any existing linter for skills that can be used?

Author: The lint part we can delay it for later stage, for now just lets just rely on manual reviews.

What the lint does: rejects any bash/sh/kubectl exec/docker/backticks in step bodies (covers "malicious PR adds bash/exec"); resolves every tool name and arg key against RegisteredToolNames() from the server package so renaming a tool auto-fails CI (covers "stale reference"); rejects literal IPs/hostnames outside an allowlist and enforces placeholders (covers "leaks customer IPs"). Plus mechanical checks: frontmatter valid, ≤ 300 lines, ends with the reporting block.

What lint can't do: prompt-injection via tool output is a runtime concern (belongs to the MCP client, not the linter); semantic drift needs PR review + layered-feature ownership. I'll drop the prompt-injection row from the lint-mitigations column to reflect that.

Existing linter? No off-the-shelf Agent Skills linter I know of; the We'd lean on markdownlint-cli2 for generic Markdown hygiene and write a small in-tree Go tool that imports the server's tool registry for the OVN-K mcp specific checks. -- but again as a stretch goal

| Prompt-injection via tool output | Skills are repo-loaded, not output-loaded; standard MCP runtime guarantee |
| Skill leaks customer IPs/hostnames | PR review; placeholders only |
| Skill drifts from feature reality | Layered-feature ownership in Phase 2; PR review by maintainers |

# Targeting the Right Pod / Node

OVN-Kubernetes is a distributed system: each node has its own NB DB,
SB DB, OVS instance, and kernel state. Every MCP tool that touches
those layers therefore needs to know *which* `ovnkube-node` pod (or
which node) to run against. Without skills, the LLM has to recompute
this mapping for every individual tool call, which is where it
typically picks the wrong pod.

Skills solve this by making "find the right pods" an explicit early
step in every workflow. For example, Step 2 of `pod-to-pod-connectivity`
is:

* `resource-list` the ovnkube pods,
* identify the `ovnkube-node` pod on the source node and the one on
  the destination node, plus an `ovnkube-control-plane` pod for NB/SB

>   queries,

Reviewer: NBDB/SBDB queries will also be done on the ovnkube-node pods. The ovnkube-control-plane pods are not used by any of the current tools.

Author: sorry this was for the downstream part not 100% specific to ovn-kubernetes, I updated it

* reuse those pod names as the `name=<pod>` parameter in every
  subsequent OVN/OVS tool call.

The MCP server itself stays stateless (per OKEP-5494); the skill is
where the "for this triage session, the source pod is X" decision is
recorded and reused.

# Deployment Strategy

1. **Bundled with the repo (Phase 1, default).** Clone the repo, point an
   Agent-Skills-compatible client at it, done.
2. **Per-client symlinks (Phase 1).** `.cursor/skills` and
   `.claude/skills` make discovery automatic.
3. **MCP resources (Phase 2).** Same skills, served over the wire to
   any MCP client - no local checkout required.

We deliberately do **not** ship a separate skills package or container
in Phase 1 - that is release-management cost without a real distribution
problem.

**Configuration.** Operators disable skills like they disable tools (per
OKEP-5494's config file): a `skills.disabled` list hides matching
`SKILL.md` from the resource provider in Phase 2; in Phase 1, simply do

> not symlink them.

Reviewer: Is this talking about diabling all the skills or selective skills?

Author: yes

# Testing Strategy

**Approach 1: replay against historical bundles (advisory).** Per-skill
fixture in `test/skills/<skill>/`: a curated must-gather / sos-report

> bundle + an `expected.md` describing the correct RCA. A runner uses the

Reviewer: How will the live-cluster mode tools be tested in this case? All the skills planned for phase-1 target only the live-cluster mode tools.

Author: Reworked the Testing Strategy. Since a skill is really a prompt, it's validated with evals rather than unit tests, but running live-cluster tools in CI is impractical (and LLM-dependent, which we don't pin), so evals are a contributor responsibility: whoever writes/changes a skill runs the eval prompts against a disposable kind cluster and pastes the summary (trigger + tool sequence + RCA + pass/fail) into the PR as review evidence. The only automated CI gate stays the static skills-lint. Offline must-gather replay is kept as an optional cheaper re-check. Framing follows the emerging eval practice for agent skills (linked the OpenAI eval-skills post). this is a good read do check it out 😄

offline-execution mode from OKEP-5494 to point the MCP server at the
bundle, runs the agent loop, and diffs the transcript against
`expected.md`. Answers two questions: did the agent pick the right skill?
did it follow it?

**Approach 2: chaos / synthetic (Phase 1).** Per-skill chaos scenarios on a
kind / production cluster (delete an LSP, blackhole UDP/53, fill the
conntrack table, install a bad drop flow, ...). A job runs the

> agent with skills enabled vs disabled and tracks the RCA-accuracy delta.

Author: modified the testing section altogether

This is the benchmark dataset that backs OKEP-5494's "LLM Capability
Assessment".

# Documentation

> OKEP on [https://ovn-kubernetes.io/](https://ovn-kubernetes.io/). End-user

Reviewer: Currently the OKEPs that are being added to this repo will not appear on the ovn-kubernetes website.

Author: should we update it somehow ?, or talk in upstream meeting ?

Reviewer: It can be discussed in the upstream meeting.

Reviewer: Let's drop this for now. For now, let's link this OKEP in the `doc/okeps/README.md` file.

> docs in `docs/skills/`:

Reviewer: Can the skills directory have a separate README.md file which can be for human documentation? The SKILL.md file will contain the info for the agents.

Author: yes, I think we should have one README.md for human documentation as
/skills/README.md

* **Getting started** - how to point Cursor / Claude Code / Cursor CLI /
  `@cursor/sdk` at the server and the skills folder.
* **Skill reference** - rendered list with frontmatter + tools used.
* **Authoring guide** - format, lint rules, review.
* **Cookbook** - 5-10 worked examples (paired with the Layer-2
  fixtures so users can replay).
* **FAQ / anti-patterns** - free-form bash, hardcoded parameters,

>   cross-cluster assumptions, etc.

Reviewer: These can reside in the `docs/skills/` directory.

Author: Either is fine skills/README.md or docs/skills/README.md. Both are directory-level so neither gets pulled into an individual skill's agent context. Only thing we avoid is a README inside skills/<name>/. 

intended to be mostly be read by humans rather than LLM's 
I can move it to docs/skills.md

**Skills as MCP resources (Phase 2).** When ready, add a resource provider
exposing each skill at e.g. `mcp://ovn-kubernetes/skills/<name>` with
`mimeType: text/markdown`. Additive; does not change the file format.

# Security Model

Skills inherit OKEP-5494's model unchanged. They neither expand nor relax
it:

* Skills only call **named MCP tools**. They cannot escape the tool
  sandbox.
* Skills are **read-only by construction** - lint rejects write verbs.
* Skills are **public**, reviewed in the open. No secrets, no IPs,
  no customer data; placeholders only (`<src-pod>`, `<node>`, `<ovn-pod>`).
* Skills carry no RBAC of their own. They run with whatever ServiceAccount
  / kubeconfig the MCP server uses.
* **Remediation** appears only as text in the Reporting block, addressed
  to the human. The agent does not execute it under this OKEP.

| Threat | Mitigation |
|---|---|
| Malicious skill PR adds bash/exec | CI lint forbids free-form shell; PR review |
| Stale skill references removed tool | CI lint cross-checks the live tool registry |
| Prompt-injection via tool output | Skills are repo-loaded, not output-loaded; standard MCP runtime guarantee |

> | Skill leaks customer IPs/hostnames | PR review; placeholders only |

Author: done

      Remediation  : Add an egress rule allowing TCP/8080 to api-back.
      Verification : Re-run ovn-trace; expect ls_in_acl allow-related.
```

The agent never invented a MAC, never opened tcpdump prematurely, and
produced a transcript a maintainer can read like a runbook execution log.

## Authoring and review process

Skills are code:

* In-tree under `skills/`, reviewed by an OVN-Kubernetes-mcp maintainer (or
  the layered-feature owner once feature skills land).
* MCP tool catalogue changes that affect skills must update both in the
  same PR; CI rejects PRs that leave a skill referencing a removed tool.

# Implementation Details

Phase-1 deliverables:

1. Author the initial set of `SKILL.md` files under canonical `skills/`
   (one directory per skill from the catalogue above). Add
   `skills/README.md` as the human-readable index + format reference.
2. Add a short docs page on pointing Cursor / Claude Code / `@cursor/sdk`
   at the server and the available skills.

In Phase-1, skill correctness is enforced by PR review (well-formedness)
plus the contributor's eval run (behaviour, see Testing Strategy). There

> is no automated skill gate yet.

Reviewer: This comment needs to be addressed.

  custom agents) without forking per-client.
* Ship an **initial set of skills** covering the most frequent
  Phase-1 triage paths: `pod-to-pod-connectivity`,
  `service-connectivity`, `network-policy-debugging`,
  `external-connectivity`, `node-networking`, `ovn-topology-discovery`,
  `ovs-flow-analysis`, `conntrack-debugging`, `packet-drop-investigation`.
* Skills MUST only call **named MCP tools** exposed by the server (no
  free-form bash), MUST be **read-only**, and MUST encode layered,
  evidence-driven workflows.
* Skills MUST live in-tree and be code-reviewed; a Phase-2 CI lint against
  the MCP tool registry adds automated drift protection.

# Future Goals

* **Feature-specific skills**: ovn-k8s skills related to features in the main product `egressip`,
  `egress-firewall`, `route-advertisements`, `ipsec`, `bgp`,
  `udn-primary-network`, `multus-secondary-networks`, etc.
* **Skills as MCP resources** - serve `SKILL.md` over the wire so remote
  MCP clients can fetch them from the server without local file access.
* **Skill benchmark suite** - per-skill fixtures (must-gather /
  sos-report bundle + expected RCA) that measure troubleshooting

>   quality over time.

Reviewer: Since the skills specifically mention usage of the live-cluster tools, this seems a little contradictory as to how the quality will be assessed using must-gather/sos-report. I would say best approach would be to add evals which will be run periodically to check the quality over time.

              +-------------------------------------------+
              |          MCP Client + LLM Agent           |
              |  (Cursor / Claude Code / @cursor/sdk)     |
              +--------+---------------------+------------+
                       | loads               | calls
                       v                     v
              +-----------------+   +---------------------+
              |   SKILL.md      |   |   MCP Tools         |
              | frontmatter +   |   | resource-get,       |
              | layered steps + |   | ovn-show, ovn-get,  |
              | decision tree   |   | ovn-trace,          |
              |                 |   | ovs-vsctl/ofctl,    |
              | references --------->  tcpdump, pwru, ... |
              | tools by name   |   |                     |
              +-----------------+   +---------------------+
                       \                  /
                        v                v
                       ovnkube-node pods, NB/SB DBs,
                       OVS, kernel, NIC
```

A skill is a deterministic recipe over the MCP tool catalogue. If the
catalogue changes, the affected skills change in the same PR -

> enforced by CI.

Reviewer 2: (thought) there may be other properties to enforce in CI, e.g. approximate token / character length of each skill, spelling and such. The official spec also suggests to use `validate` verb from https://github.com/agentskills/agentskills/tree/main/skills-ref

read or run so I learn the canonical debug flow without shoulder-surfing
a senior engineer.

**As a triage engineer**, I want the agent to fetch parameters from earlier
tool output (MACs, port names, datapath names) instead of asking me to
type them.

**As a layered-feature owner** (EgressIP, IPSec, BGP, RouteAdvertisements,
UDN, ...), I want to publish a skill for my feature's failure modes so the
LLM does not have to rediscover them.

# Proposed Solution

## Repository layout

```text
ovn-kubernetes-mcp/
  skills/                              <- canonical, vendor-neutral
    README.md                          <- human-readable index + format
                                          reference (agents load per-skill
                                          SKILL.md, not this file)
    pod-to-pod-connectivity/SKILL.md
    service-connectivity/SKILL.md
    ...
```

A single canonical `skills/` is the source of truth (and what we will
serve as MCP resources later). Each client is pointed at this directory
through its own skills configuration.

## Skill format

```markdown
---
name: <kebab-case-skill-name>             # matches directory name
description: <symptom-oriented; must

>   contain "Use when ..." trigger phrases>

Reviewer 2: would `allowed-tools` be useful here? Also, maybe we should add `license: Apache-2.0` if there's a plan to serve the documents via API later?

Reviewer 2: No I guess. It's fine, just wanted to mention that if you serve Apache material, you have to distribute it with license text (which, if you squint enough, can probably be substituted by a frontmatter tag).

---

# <Title>

## Prerequisites
- inputs to collect from the user

> ## Tools used

Reviewer 2: The spec defines `allowed-tools` frontmatter field. Is it different?

  - resource-get both pods                -> podIP, nodeName, Running
  - resource-list ovnkube pods            -> picks src/dst ovnkube-node
                                             (each hosts its own NB/SB/OVS
                                             in OVN-IC mode)
  - ovn-show NB; ovn-get LSP for both;
    ovn-get Port_Binding; ovn-get Chassis -> LSPs up=true, ports bound
  - resource-list NetworkPolicy ns=shop   -> finds default-deny-egress
  - ovn-trace with eth/ip from above      -> drop in ls_in_acl
  - decision tree -> hops to network-policy-debugging
                  -> identifies the offending ACL
  - Reports:
      Root cause   : NetworkPolicy 'default-deny-egress' selects
                     curl-front; matching allow-rule is missing.
      Evidence     : ovn-trace stage ls_in_acl, ACL UUID xxx,
                     external_ids k8s.ovn.org/name=default-deny-egress.
      Remediation  : Add an egress rule allowing TCP/8080 to api-back.
      Verification : Re-run ovn-trace; expect ls_in_acl allow-related.
```

The agent never invented a MAC, never opened tcpdump prematurely, and
produced a transcript a maintainer can read like a runbook execution log.

> ## Authoring and review process

Reviewer 2: I am new to this field, so this may be obvious to everyone but me, but: how does one review a change to an existing skill? Of course, if there's an obvious bug in logic, we can review and edit it like we do any other type of documentation. What does one do with PRs that reshuffle / adjust wording of particular skill steps while claiming it makes an agent operate better / more efficiently? How does this claim get independently confirmed? What are the quality properties that define a good and a bad skill? Can these be quantified?

Skills are code:

* In-tree under `skills/`, reviewed by an OVN-Kubernetes-mcp maintainer (or
  the layered-feature owner once feature skills land).
* MCP tool catalogue changes that affect skills must update both in the
  same PR; CI rejects PRs that leave a skill referencing a removed tool.

# Implementation Details

Phase-1 deliverables:

1. Author the initial set of `SKILL.md` files under canonical `skills/`
   (one directory per skill from the catalogue above). Add
   `skills/README.md` as the human-readable index + format reference.

> 2. Add a short docs page on pointing Cursor / Claude Code / `@cursor/sdk`

Reviewer 2: Why these three but not others? I don't see Codex here, but I also don't see any truly open harnesses like opencode or pi.

I'm not suggesting to describe these though. I think skills are meant to be harness agnostic, not sure docs should go into configuring particular agents to use them. Agent space is moving rapidly with new fads every week, so I'd rather avoid the hassle completely.

Reviewer 2: I mean, the less work to do the better in my mind. Up to you.

Author: done

   at the server and the available skills.

In Phase-1, skill correctness is enforced by PR review (well-formedness)
plus the contributor's eval run (behaviour, see Testing Strategy). There
is no automated skill gate yet.

Phase-2 deliverables:

> 1. **`skills-lint` (`make skills-lint`, wired into CI).** A static checker

Reviewer 2: I would expect phase-1 to include a basic static linter. See https://github.com/agentskills/agentskills/tree/main/skills-ref as example.

**Configuration.** In Phase-2, operators selectively disable individual
skills via a `skills.disabled` list in the MCP server config, which hides
matching skills from the resource provider. Phase-1 has no config-driven toggle: a client is pointed at the
`skills/` directory as a whole (directory-level, all-or-nothing), so
selectively disabling one skill means omitting its directory from
`skills/`, and disabling everything means not pointing the client at
`skills/` at all.

# Testing Strategy

A skill is fundamentally a prompt for an agent, so it is validated the
way prompts are - with **evals**, not conventional unit tests. This
follows the emerging practice for agent skills in the wider ecosystem
(see OpenAI's
[Testing Agent Skills Systematically with Evals](https://developers.openai.com/blog/eval-skills)).
An eval is a simple loop:

`symptom prompt → captured agent run (tool-call trace + final RCA) → checks → score`

## Who runs evals, and when

Evals are **run by the contributor** who authors or changes a skill, not
by CI. Creating an OVN-Kubernetes cluster and connecting the MCP server with the live-cluster tools (`ovn-trace`, `ovs-appctl ofproto/trace`, `tcpdump`, `pwru`, ...)

> is impractical to run as an automated gate, and the result depends on the

Reviewer 2: I am not sure I follow the logic here. Yes, we don't pin the model, but neither the author does. Nor their harness / system prompts are. So if we were to have a choice, I'd rather rely on a CI signal confirming a skill "works" in a set of modeled situations than let the author claim, without a proof, that it's an improvement.

This is not a trivial matter to set evals and scoring that is (kinda) deterministic, and maybe we don't want to deal with it - because it's hard to set it up, or because it means inference costs no one is covering. But let's be honest here: an author saying their PR is an improvement cannot be taken as a proof, one way or another, and the project owners will have to judge / eval for themselves one way or another.

Reviewer 2: I understand the constraints. I just don't see how...

```
any reviewer can re-run the same thing on their side
```

...since a reviewer 1) doesn't have access to their cluster in their broken state; 2) doesn't have tools (CI or manual) to replicate _any_ similarly broken state with enough fidelity to prove quality of the contribution; 3) doesn't have a regression suite to prove nothing major is broken by a proposed change - or to add more regression cases to as new breakages happen.

In lack of these, as an imaginary reviewer of a skill change, I don't see how I can decide whether a contribution should be accepted or not. It's all vibes at this point.

Personally, I wouldn't want to be a reviewer on a project with no means to test changes. YMMV.

This is a non-blocking comment btw (most here are not, they are shared as food for thought).

Author: agree, added.

backing LLM, which we do not pin. Instead:

* The author runs the skill's eval prompts against a cluster they have

>   access to (a disposable kind cluster is sufficient or a production

Reviewer 2: Considering these are "debugging" skills, are we also to define specific *states* of aforementioned clusters-under-test that model the skill target? What's the format in which these states are defined?

Reviewer 2: It's encouraging that you seem to agree that some library of "broken states" (imperative scripts that deterministically break clusters) to test skills against would be valuable. I like it.

If I understand methodology, one would...

1. deploy a cluster in known healthy state
2. for each "breaking script"...
2.1. run it against the cluster, putting it in a known broken state
2.2. run a (reasonably pinned) harness system (llm, prompt, agent version) against this cluster
2.3. assess the report from 2.2 against an "expected" checklist

(2.3) can then be a "human comparing the notes" or a LLM judge doing it.

I think this is a good direction and I'm looking forward to see how this pans out in code.

Author: > deploy a cluster in known healthy state
> for each "breaking script"...
> 2.1. run it against the cluster, putting it in a known broken state
> 2.2. run a (reasonably pinned) harness system (llm, prompt, agent version) against this cluster
> 2.3. assess the report from 2.2 against an "expected" checklist
> (2.3) can then be a "human comparing the notes" or a LLM judge doing it.

yes, this is a reasonable and good approach to start off with.
and the broken state might not be an exhaustive list of all the issues you can think of but more of an exmaple set.
should I add a line to the okep capturing that author should share the script of what he used to test the skill against, or is it implicitly understood.

Author: Added a small line to state this explicitly

  test output, alongside reading the skill itself.

In Phase-1 there is no automated skill gate - well-formedness is checked
in PR review and behaviour by the contributor's eval run. Phase-2 adds the
static `skills-lint` (Implementation Details) as an automated CI gate for
well-formedness (frontmatter, tool-reference resolution, no free-form
bash, reporting block, line cap). Either way, review / lint validates that
a skill is *well-formed*; the contributor's eval run validates that it
*works*.

## What an eval checks

Following the standard eval taxonomy, a skill is graded on four axes:

* **Triggering** - given a symptom prompt, did the correct skill get
  selected from its `description`? Include **negative controls** (prompts
  that must NOT trigger it) to catch over-eager matching.
* **Process** - did the agent follow the skill's layered sequence,
  calling the expected tools on the expected pods/nodes?
* **Outcome** - did the final RCA name the correct root cause and cite the
  tool output that proves it?
* **Efficiency** - did it get there without thrashing (tool-call count,
  token use) versus running without the skill?

## Eval prompt set

> Each skill ships with a small (~10-20 prompt) set of symptom phrasings -

Reviewer 2: Practically, is it some kind of yaml with instructions?

Reviewer 2: I think we should indeed capture smth like your example and then let a "judge" of the triage report (whether a human or a machine) to assess if all checkboxes were ticked.

Author: agree do you want to add a small line about this in OKEP as an agreement ?

explicit ("use the pod-to-pod-connectivity skill"), implicit ("pods on
node A can't reach node B"), and negative controls - kept next to the
skill. It grows over time: every real miss found in use becomes a new
prompt, so the set becomes a living regression record the next
contributor re-runs.

## Optional: offline replay

> Where a scenario has a captured `must-gather` / `sos-report` bundle, a

Reviewer 2: Maybe worth to build a library of "breaking scenarios" - deterministic steps that can be executed against a cluster-under-test (kind) that put it in a known breaking state that a particular skill is supposed to target.

  review. Out of scope until those land.

# Non-Goals

* **No new format.** We adopt Agent Skills exactly. Inventing our own
  defeats portability across MCP/agent runtimes.
* **We do not own client loading.** How Cursor or Claude Code finds a
  `SKILL.md` is the client's concern.
* **Not a tool replacement.** Skills are guidance; the tools live in the
  MCP server. Without the server, a skill is just a runbook.
* **No write / remediation.** Phase 1 stays read-only.
* **Not solving LLM quality.** A bad model + a good skill still fails;
  skills bound the failure mode, they do not eliminate it.

# Introduction

## Why skills, not just tool descriptions or a system prompt?

Tool descriptions cannot encode:

1. **Methodology.** "Always check the pod is Running before running
   `ovn-trace`" is a workflow rule, not a property of any single tool.
2. **Cross-tool wiring.** `addresses` from `ovn-get Logical_Switch_Port`
   becomes the `eth.src`/`ip4.src` of `ovn-trace`. That graph lives
   between tools, not inside any one of them.
3. **Decision trees.** "If `ovn-trace` allows but `ofproto-trace`
   drops, suspect flow-installation; restart `ovn-controller`" is

>    exactly what skills are for.

Author: Fixed (line 112-114): ofproto-trace → ovs-appctl ofproto/trace (the canonical tool + action, matching the OVS tool consolidation we applied earlier). Troubleshooting guidance preserved verbatim — only the tool name changed. No ofproto-trace occurrences remain.

type them.

**As a layered-feature owner** (EgressIP, IPSec, BGP, RouteAdvertisements,
UDN, ...), I want to publish a skill for my feature's failure modes so the
LLM does not have to rediscover them.

# Proposed Solution

## Repository layout

```text
ovn-kubernetes-mcp/
  skills/                              <- canonical, vendor-neutral
    README.md                          <- human-readable index + format
                                          reference (agents load per-skill
                                          SKILL.md, not this file)
    pod-to-pod-connectivity/SKILL.md
    service-connectivity/SKILL.md
    ...
```

A single canonical `skills/` is the source of truth (and what we will
serve as MCP resources later). Each client is pointed at this directory
through its own skills configuration.

## Skill format

```markdown
---
name: <kebab-case-skill-name>             # matches directory name
description: <symptom-oriented; must
  contain "Use when ..." trigger phrases>
---

# <Title>

## Prerequisites
- inputs to collect from the user

## Step 1 ... Step N
<imperative title>
<MCP tool call: tool name + parameters>

> <what to extract; which fields feed which later step>

Author: The SKILL.md body is free-form Markdown per the [Agent Skills spec](https://agentskills.io/specification) — there's no standard machine-readable step grammar, and pinning one in the OKEP would over-specify an intro proposal. The Phase-2 skills-lint can validate tool/parameter references by scanning for the tool names against the server registry without us mandating a strict call syntax here. I'd rather leave the template illustrative and let the lint implementation decide the exact parse rules. Open to revisiting if we hit ambiguity in practice.

## Decision Tree
<branching summary; may hop to a sibling skill>

## Reporting Findings
1. Root cause   2. Evidence   3. Remediation   4. Verification
```

**Author rules** (CI-enforced):

1. Frontmatter present; `name` matches directory; `description` non-empty
   and contains "Use when ..." phrasing.
2. Every tool reference resolves to a registered MCP tool, every
   parameter is a real parameter of that tool.
3. No free-form bash, ssh, `kubectl exec`, or write verbs (`set`, `add`,
   `del`, `remove`, ...).

> 4. Ends with the four-bullet reporting block.

Author: it will be referring to the existing tools itself, and all the tools currently are read only in the codebase

> 4. Ends with the four-bullet reporting block.

Author: scoped for next phase

| Skill leaks customer IPs/hostnames | PR review; placeholders only |
| Skill drifts from feature reality | Layered-feature ownership in Phase 2; PR review by maintainers |

# Targeting the Right Pod / Node

OVN-Kubernetes is a distributed system: each node has its own NB DB,
SB DB, OVS instance, and kernel state. Every MCP tool that touches
those layers therefore needs to know *which* `ovnkube-node` pod (or
which node) to run against. Without skills, the LLM has to recompute
this mapping for every individual tool call, which is where it
typically picks the wrong pod.

Skills solve this by making "find the right pods" an explicit early
step in every workflow. For example, Step 2 of `pod-to-pod-connectivity`
is:

* `resource-list` the ovnkube pods,
* identify the `ovnkube-node` pod on the source node and the one on
  the destination node - in OVN-IC mode each of these pods hosts its own
  local NB/SB databases and OVS instance, so every OVN/OVS query for that
  node routes through the same pod,
* reuse those pod names as the `name=<pod>` parameter in every
  subsequent OVN/OVS tool call.

The MCP server itself stays stateless; the skill is where the "for this

> triage session, the source pod is X" decision is recorded and reused.

Author: Skipped — no longer applicable. removed section

recorded and reused.

# Deployment Strategy

1. **Bundled with the repo (Phase 1, default).** Clone the repo, point an
   Agent-Skills-compatible client at it, done.
2. **Per-client symlinks (Phase 1).** `.cursor/skills` and
   `.claude/skills` make discovery automatic.
3. **MCP resources (Phase 2).** Same skills, served over the wire to
   any MCP client - no local checkout required.

We deliberately do **not** ship a separate skills package or container
in Phase 1 - that is release-management cost without a real distribution
problem.

**Configuration.** Operators disable skills like they disable tools (per
OKEP-5494's config file): a `skills.disabled` list hides matching
`SKILL.md` from the resource provider in Phase 2; in Phase 1, simply do

> not symlink them.

Author: we already did

> not symlink them.

Author: that's expected

# Testing Strategy

A skill is fundamentally a prompt for an agent, so it is validated the
way prompts are - with **evals**, not conventional unit tests. This
follows the emerging practice for agent skills in the wider ecosystem
(see OpenAI's
[Testing Agent Skills Systematically with Evals](https://developers.openai.com/blog/eval-skills)).
An eval is a simple loop:

`symptom prompt → captured agent run (tool-call trace + final RCA) → checks → score`

## Who runs evals, and when

Evals are **run by the contributor** who authors or changes a skill, not
by CI. Standing up an OVN-Kubernetes cluster and driving live-cluster
tools (`ovn-trace`, `ovs-appctl-ofproto-trace`, `tcpdump`, `pwru`, ...)
is impractical to run as an automated gate, and the result depends on the
backing LLM, which we do not pin. Instead:

* The author runs the skill's eval prompts against a cluster they have
  access to (a disposable kind cluster is sufficient or a production

>   cluster) and **pastes the eval summary into the PR**: which prompts

Author: This shouldn't matter, as the tests are done by the human, so as long as he provides enough evidence it should be enough, environment shouldn't matter

>   cluster) and **pastes the eval summary into the PR**: which prompts

Author: please ignore this, this is out of scope of the repo, user can use what ever cluster as long as they can present reasonable input and outputs

# OKEP-59: Network Troubleshooting Skills for OVN-Kubernetes

* Issue: [#59](https://github.com/ovn-kubernetes/ovn-kubernetes-mcp/issues/59)

# Problem Statement

The `ovn-kubernetes-mcp` server today exposes read-only debugging tools
across 5 layers - Kubernetes API (`resource-get`, `resource-list`,
`pod-logs`), OVN NB/SB (`ovn-show`, `ovn-get`, `ovn-lflow-list`,
`ovn-trace`), OVS (`ovs-vsctl`, `ovs-ofctl`, `ovs-appctl` - each takes an
`action` subcommand, e.g. `ovs-vsctl` action `show`/`list-br`/`list-ports`,
`ovs-ofctl` action `dump-flows`, `ovs-appctl` action
`dpctl/dump-conntrack`/`ofproto/trace`), kernel networking
(`get-conntrack`, `get-iptables`, `get-nft`, `get-ip`), and packet
capture / tracing (`tcpdump`, `pwru`). Those tools are the *inputs* to a
triage; they are not the triage itself. What is missing today is any
guidance for the LLM on how to compose them into a deterministic, layered
troubleshooting workflow.

Without that guidance, LLMs misuse the tool surface in predictable

> ways:

Reviewer: Is it predicatable or unpredictable?

Author: I reworded it to unpredicatable  😄 , but thinking about it LLMs usually do it in predicatable ways ( with the given tools ), but with fable 5 and etc, I can't really say they would do it unpredicatable ways anymore

but skills are the best way to teach how to use them, in any particular scenario, now a days these agent harnesses ( cursor, codex, claude code ) -- they are not even loading whole MCPs into each request's context, they are building something called as memory graph ( which is sort of its mind map ) with tool search, to give LLM the right tool for a required cause

layered order.

Skills add a layer of **determinism** to an otherwise non-deterministic
agent loop: the *which tools, in what order, with which parameters
threaded between them* part of triage is fixed by the skill, leaving the
LLM responsible only for symptom matching, parameter extraction from
user prompts, and final analysis. The same bug investigated twice
should produce substantially the same transcript, which makes
reproducibility, review, and regression testing tractable.

# Goals

* Define a **skill format** and **skill catalogue** for OVN-Kubernetes
  network triage shipped alongside the MCP server.
* Adopt the cross-vendor [Agent Skills](https://www.anthropic.com/news/skills)
  convention (`SKILL.md` with `name` + `description` frontmatter) so skills
  work in any compatible runtime (Cursor, Cursor CLI, Claude Code,
  Anthropic Skills API, custom agents) without forking
  per-client.
* Ship an **initial set of skills** covering the most frequent
  Phase-1 triage paths: `pod-to-pod-connectivity`,
  `service-connectivity`, `network-policy-debugging`,
  `external-connectivity`, `node-networking`, `ovn-topology-discovery`,

>   `ovs-flow-analysis`, `conntrack-debugging`, `packet-drop-investigation`.

Author: DNS is not mentioned anywhere in the OKEP

>   `ovs-flow-analysis`, `conntrack-debugging`, `packet-drop-investigation`.

Reviewer: I think the issue #59 should be updated to not mention DNS.

>   `ovs-flow-analysis`, `conntrack-debugging`, `packet-drop-investigation`.

Reviewer: The commit message also mentions DNS.

* Skills MUST only call **named MCP tools** exposed by the server (no
  free-form bash), MUST be **read-only**, and MUST encode layered,
  evidence-driven workflows.
* Skills MUST live in-tree and be code-reviewed; a Phase-2 CI lint against

>   the MCP tool registry adds automated drift protection.

Author: Doesn't apply now, lets punt this to phase- 2

2. **Cross-tool wiring.** `addresses` from `ovn-get Logical_Switch_Port`
   becomes the `eth.src`/`ip4.src` of `ovn-trace`. That graph lives
   between tools, not inside any one of them.
3. **Decision trees.** "If `ovn-trace` allows but `ofproto-trace`
   drops, suspect flow-installation; restart `ovn-controller`" is
   exactly what skills are for.
4. **Context economics.** A monolithic system prompt covering all N
   scenarios is in context for every conversation. Agent Skills load
   only the matching skill body via *progressive disclosure*: the
   `description` is always loaded; the body is loaded only when the
   user's symptom matches.

## Two-layer contract

```text
              +-------------------------------------------+
              |          MCP Client + LLM Agent           |
              |  (Cursor / Claude Code / @cursor/sdk)     |
              +--------+---------------------+------------+
                       | loads               | calls
                       v                     v
              +-----------------+   +---------------------+
              |   SKILL.md      |   |   MCP Tools         |

>               |   (this OKEP)   |   |   (the server)      |

Reviewer: Let's remove this line itself.

Author: done

| `ovn-topology-discovery`    | Map / understand the logical network | ovn-show NB/SB, ovn-get LS/LR/LSP/LRP/Chassis/Port_Binding/NAT/Static_Route/Load_Balancer, ovs-vsctl show |
| `ovs-flow-analysis`         | Logical vs OpenFlow mismatch, stale flows | ovn-lflow-list, ovs-ofctl dump-flows, ovs-appctl ofproto/trace, ovn-trace, pod-logs ovn-controller |
| `conntrack-debugging`       | NAT broken, stale entries, table-full, zone conflicts | ovs-appctl dpctl/dump-conntrack, get-conntrack -L/-S/-C, ovs-ofctl dump-flows ct, ovn-lflow-list ct |
| `packet-drop-investigation` | Silent drops, unknown drop point | tcpdump, ovs-ofctl dump-flows drop, ovs-appctl ofproto/trace, ovn-trace, ovs-appctl dpctl/dump-conntrack, get-iptables, get-nft, pwru |

Skills are independent (pick one by symptom) and **composable** - a skill
**may** hand off to a sibling when the evidence points elsewhere. This
happens two ways: a skill can explicitly reference a sibling as a next
step, or, mid-execution, the agent may notice the symptoms now match a
different skill (for example, while running `network-policy-debugging` it
finds the traffic is blocked by a node firewall, not an ACL, and moves to

> `external-connectivity`). Which sibling, if any, is loaded depends on what

Reviewer: Since the agent can hop to a different skill based on the current evidence, how do we ensure that it does not keep on thrashing between same skills and end up in a loop?

Author: Hops are evidence-driven, not symptom-driven, the agent only moves on after a tool call rules the current skill out, so it's converging, not oscillating. Tool outputs accumulate in context and all tools are read-only, so re-running one yields the same answer it already has, no reason to repeat.  hard termination is the agent runtime's turn/iteration limit, same as any tool-use loop.

Reviewer: I am not completely sure of this, but this is probably out-of-scope of the skills. Only thing we can do is probably provide a general entrypoint which mentions that for diagnosing one issue, a skill can be invoked only once.

> the evidence shows - it is not a fixed dependency between skills.

Author: its upto the agent, for example it can handoff at the end of skill that it finds it more of a node level issue or something, or at the mid-way there are clear hints it can't be that particular issue

# User Stories

> **As an OVN-Kubernetes developer**, I want the AI assistant to follow the
> layered triage methodology I would (pod state → NB → SB → OVS → kernel →
> capture) so I get a consistent, auditable RCA instead of a model
> improvising from training data.

Reviewer: nit
```suggestion
**As an OVN-Kubernetes developer**, I want the AI assistant to follow the
layered triage methodology that I would follow (pod state → NB → SB → OVS → kernel →
capture) so I get a consistent, auditable RCA instead of a model
improvising from training data.
```

Author: done

**As a new engineer joining the team**, I want a published skill I can
read or run so I learn the canonical debug flow without shoulder-surfing
a senior engineer.

**As a triage engineer**, I want the agent to fetch parameters from earlier
tool output (MACs, port names, datapath names) instead of asking me to
type them.

**As a layered-feature owner** (EgressIP, IPSec, BGP, RouteAdvertisements,
UDN, ...), I want to publish a skill for my feature's failure modes so the
LLM does not have to rediscover them.

# Proposed Solution

## Repository layout

```text
ovn-kubernetes-mcp/
  skills/                              <- canonical, vendor-neutral
    README.md                          <- human-readable index + format
                                          reference (agents load per-skill
                                          SKILL.md, not this file)
    pod-to-pod-connectivity/SKILL.md
    service-connectivity/SKILL.md
    ...
```

A single canonical `skills/` is the source of truth (and what we will
serve as MCP resources later). Each client is pointed at this directory
through its own skills configuration.

## Skill format

```markdown
---

> name: <kebab-case-skill-name>             # matches directory

Reviewer: nit
```suggestion
name: <kebab-case-skill-name>             # matches directory name
```

Author: done

description: <symptom-oriented; must
  contain "Use when ..." trigger phrases>
---

# <Title>

## Prerequisites
- inputs to collect from the user

## Step 1 ... Step N
<imperative title>
<MCP tool call: tool name + parameters>
<what to extract; which fields feed which later step>

## Decision Tree
<branching summary; may hop to a sibling skill>

## Reporting Findings
1. Root cause   2. Evidence   3. Remediation   4. Verification

> ```

Author: skills lint will decided in phase - 2, I will get 100% confidence only while implementing it

**Author rules** (CI-enforced):

1. Frontmatter present; `name` matches directory; `description` non-empty
   and contains "Use when ..." phrasing.
2. Every tool reference resolves to a registered MCP tool, every
   parameter is a real parameter of that tool.
3. No free-form bash, ssh, `kubectl exec`, or write verbs (`set`, `add`,
   `del`, `remove`, ...).
4. Ends with the four-bullet reporting block.

> 5. ≤ 500 lines. Cross-reference sibling skills instead of duplicating.

Reviewer: I want to understand how the above mentioned checks will be enforced. This need not be detailed in the OKEP itself. But want to ensure that what is being mentioned in the OKEP can be actually achieved or not.

Author: tweaked a little how we can enforce this going forward  in phase - 2

ok I found how some open source projects use this
1 - [skills-ref](https://github.com/agentskills/agentskills/tree/main/skills-ref)The official reference validator from agentskills.io. skills-ref validate path/to/skill

2 - [skill-doctor](https://github.com/tcarac/skill-doctor)

can they achieve all that is mentioned -- No I guess

just to listed points above
1 - Frontmatter present, valid YAML, name matches dir, description non-empty -- yes they can do this
2 - still human review is required here
3 - this is what we want to achieve -- so still going to be human review
4 - easy to check
5 - Not standard, but trivial to add

I can check and let you know how other opensource projects are taking this forward, and what they have as guidance.

## How the agent uses a skill

1. Indexes all `SKILL.md`; only frontmatter is loaded by default.
2. On a user prompt, matches against `description` fields; loads the
   matching body.
3. Follows the steps, calling MCP tools with parameters extracted from
   the user prompt and prior tool output.
4. At each decision rule, continues, hops to a sibling skill, or
   returns the reporting block.

We do not extend MCP, do not require a new client feature, and do not
fork existing clients. Works today in Cursor, Cursor CLI, Claude Code,
Anthropic Skills API, and custom agents using `@cursor/sdk`.

## Worked example - "Pod A can't reach Pod B"

```text
User: "curl-front-7c4 (ns shop) and api-back-9d2 (ns shop) can't talk
       on TCP/8080. Used to work."

Agent:

>   - matches symptom -> loads pod-to-pod-connectivity

Reviewer: Doesn't the `network-policy-debugging` skill also match this?

Author: thinking hard on this if you want phase - 1 to have just non-overlapping symptoms

Rethinking the catalogue: (pod-to-pod-connectivity, service-connectivity, external-connectivity, node-networking). The others (network-policy-debugging, ovs-flow-analysis, conntrack-debugging, packet-drop-investigation, ovn-topology-discovery) are techniques can be used within those triages, not separate symptoms  or may be we can punt them to phase 2 or remove altogether if and revist later 

suggest if you can think anything more can be added, and think about this opinion

Reviewer: Let's just drop the others for now. We can revisit them once we get the first set of skills working correctly along with evals.

Reviewer 2: I agree this can start with a handful of obvious ones; iterate later adding new / refactoring.

  - resource-list ovnkube pods            -> picks src/dst ovnkube-node
                                             (each hosts its own NB/SB/OVS
                                             in OVN-IC mode)
  - ovn-show NB; ovn-get LSP for both;
    ovn-get Port_Binding; ovn-get Chassis -> LSPs up=true, ports bound
  - resource-list NetworkPolicy ns=shop   -> finds default-deny-egress
  - ovn-trace with eth/ip from above      -> drop in ls_in_acl
  - decision tree -> hops to network-policy-debugging
                  -> identifies the offending ACL
  - Reports:
      Root cause   : NetworkPolicy 'default-deny-egress' selects
                     curl-front; matching allow-rule is missing.
      Evidence     : ovn-trace stage ls_in_acl, ACL UUID xxx,
                     external_ids k8s.ovn.org/name=default-deny-egress.
      Remediation  : Add an egress rule allowing TCP/8080 to api-back.
      Verification : Re-run ovn-trace; expect ls_in_acl allow-related.
```

The agent never invented a MAC, never opened tcpdump prematurely, and
produced a transcript a maintainer can read like a runbook execution log.

## Authoring and review process

Skills are code:

> * In-tree under `skills/`, reviewed by an OVN-Kubernetes maintainer (or

Reviewer: nit
```suggestion
* In-tree under `skills/`, reviewed by an OVN-Kubernetes-MCP maintainer (or
```

Author: done

| Skill leaks customer IPs/hostnames | PR review; placeholders only |
| Skill drifts from feature reality | Layered-feature ownership in Phase 2; PR review by maintainers |

# Targeting the Right Pod / Node

OVN-Kubernetes is a distributed system: each node has its own NB DB,
SB DB, OVS instance, and kernel state. Every MCP tool that touches
those layers therefore needs to know *which* `ovnkube-node` pod (or
which node) to run against. Without skills, the LLM has to recompute
this mapping for every individual tool call, which is where it
typically picks the wrong pod.

Skills solve this by making "find the right pods" an explicit early
step in every workflow. For example, Step 2 of `pod-to-pod-connectivity`
is:

* `resource-list` the ovnkube pods,
* identify the `ovnkube-node` pod on the source node and the one on
  the destination node - in OVN-IC mode each of these pods hosts its own
  local NB/SB databases and OVS instance, so every OVN/OVS query for that
  node routes through the same pod,
* reuse those pod names as the `name=<pod>` parameter in every
  subsequent OVN/OVS tool call.

The MCP server itself stays stateless; the skill is where the "for this

> triage session, the source pod is X" decision is recorded and reused.

Reviewer: To be honest, the current agents do this already without the skills as well. I don't think we need to mention this explicitly in the OKEP.

Author: agreed removed this section

We deliberately do **not** ship a separate skills package or container
in Phase 1 - that is release-management cost without a real distribution
problem.

**Configuration.** In Phase-2, operators selectively disable individual
skills via a `skills.disabled` list in the MCP server config, which hides
matching skills from the resource provider. Phase-1 has no config-driven toggle: a client is pointed at the
`skills/` directory as a whole (directory-level, all-or-nothing), so
selectively disabling one skill means omitting its directory from
`skills/`, and disabling everything means not pointing the client at
`skills/` at all.

# Testing Strategy

A skill is fundamentally a prompt for an agent, so it is validated the
way prompts are - with **evals**, not conventional unit tests. This
follows the emerging practice for agent skills in the wider ecosystem
(see OpenAI's
[Testing Agent Skills Systematically with Evals](https://developers.openai.com/blog/eval-skills)).
An eval is a simple loop:

`symptom prompt → captured agent run (tool-call trace + final RCA) → checks → score`

## Who runs evals, and when

Evals are **run by the contributor** who authors or changes a skill, not

> by CI. Standing up an OVN-Kubernetes cluster and driving live-cluster

Reviewer: nit
```suggestion
by CI. Creating an OVN-Kubernetes cluster and connecting the MCP server with the live-cluster
```

Author: done

no cluster - a cheaper way to re-check the read-only steps of a skill.
This is a convenience, not a requirement.

# Documentation

OKEP on [https://ovn-kubernetes.io/](https://ovn-kubernetes.io/). End-user
docs in `docs/skills/`:

* **Getting started** - how to point Cursor / Claude Code / Cursor CLI /
  `@cursor/sdk` at the server and the skills folder.
* **Skill reference** - rendered list with frontmatter + tools used.
* **Authoring guide** - format, lint rules, review.
* **Cookbook** - 5-10 worked examples (paired with the eval prompt sets
  so users can replay).
* **FAQ / anti-patterns** - free-form bash, hardcoded parameters,
  cross-cluster assumptions, etc.

# Known Risks and Limitations

* **Client-coverage drift.** Agent Skills is real today in Cursor,
  Cursor CLI, Claude Code, and Anthropic's Skills API, but not yet
  universal. Clients that ignore `SKILL.md` get the server's raw tool
  catalogue and the failure modes that motivated this OKEP. Mitigation:
  Phase 2 serves skills via MCP resources, the most portable channel.
* **LLM still in the loop.** Skills constrain the workflow but do not
  eliminate hallucination. The reporting block forces evidence

>   citations, which makes review easier but does not prevent errors.

Author: I think with LLM this assumption is given, LLM's will hallucinate at times

>   citations, which makes review easier but does not prevent errors.

Author: Already updated — the text now reads "the reporting block asks the agent to provide evidence citations, which makes review easier but does not prevent errors (nothing validates the citations themselves)." That matches your point: it requires the agent to supply them, it doesn't enforce/verify them. A validator would be a future add.

# OKEP-59: Network Troubleshooting Skills for OVN-Kubernetes

* Issue: [#59](https://github.com/ovn-kubernetes/ovn-kubernetes-mcp/issues/59)

# Problem Statement

The `ovn-kubernetes-mcp` server today exposes ~20 read-only debugging
tools across 5 layers - Kubernetes API (`resource-get`, `resource-list`,
`pod-logs`), OVN NB/SB (`ovn-show`, `ovn-get`, `ovn-lflow-list`,
`ovn-trace`), OVS (`ovs-vsctl-show`, `ovs-list-br`, `ovs-list-ports`,
`ovs-list-ifaces`, `ovs-ofctl-dump-flows`, `ovs-appctl-dump-conntrack`,

> `ovs-appctl-ofproto-trace`), kernel networking (`get-conntrack`,

Reviewer: The ovs tools have been changed. Please update them to the current set of OVS layer tools.

Author: Done

> `get-iptables`, `get-nft`, `get-ip`), and packet capture (`tcpdump`).

Reviewer: Please add `pwru` as well.

Author: Done

* Missing tribal knowledge ("a NetworkPolicy that selects a pod denies
  *all* traffic in that direction unless explicitly allowed").

This is not a model problem alone; it is a guidance problem. Engineers who
debug OVN-Kubernetes follow a layered, evidence-driven approach that
nobody has codified. New team members and the LLM both have to rediscover
it.

This OKEP introduces **Network Troubleshooting Skills**: short,
version-controlled markdown playbooks that the agent loads
on demand and follows step-by-step, calling MCP tools in a deterministic
layered order.

Skills add a layer of **determinism** to an otherwise non-deterministic
agent loop: the *which tools, in what order, with which parameters
threaded between them* part of triage is fixed by the skill, leaving the
LLM responsible only for symptom matching, parameter extraction from
user prompts, and final analysis. The same bug investigated twice
should produce substantially the same transcript, which makes
reproducibility, review, and regression testing tractable.

# Goals

* Define a **skill format** and **skill catalogue** for OVN-Kubernetes
  network triage shipped alongside the MCP server.

> * Adopt the cross-vendor [Agent Skills](https://www.anthropic.com/news/skills)

Reviewer: Add reference to https://agentskills.io/specification which contains the actual specification. Can any other fields the frontmatter be included apart from name and description?

Author: I mean there are other feilds, but I haven't seen anyone using them in production
there are the other fields, but I highly doubt these are required for us.

Field | Required | Constraints
-- | -- | --
name | Yes | Max 64 characters. Lowercase letters, numbers, and hyphens only. Must not start or end with a hyphen.
description | Yes | Max 1024 characters. Non-empty. Describes what the skill does and when to use it.
license | No | License name or reference to a bundled license file.
compatibility | No | Max 500 characters. Indicates environment requirements (intended product, system packages, network access, etc.).
metadata | No | Arbitrary key-value mapping for additional metadata.
allowed-tools | No | Space-separated string of pre-approved tools the skill may use. (Experimental)

what do you mean by reference here ? and  example ?.
that link is the actual specification, which is started by the anthropic

Reviewer: You have added reference to https://www.anthropic.com/news/skills and not the actual skills specification here.

`allowed-tools` could be useful, but since it's experimental, we can skip it for now. We'll have to revisit this later on.

Author: done check now

  convention (`SKILL.md` with `name` + `description` frontmatter) so skills
  work in any compatible runtime (Cursor, Cursor CLI, Claude Code,
  Anthropic Skills API, custom agents) without forking
  per-client.

> * Ship an **initial set of few skills** covering the most frequent

Reviewer: nit:

**initial set of  skills**

Author: done

  Phase-1 triage paths: `pod-to-pod-connectivity`,
  `service-connectivity`, `network-policy-debugging`,
  `external-connectivity`, `node-networking`, `ovn-topology-discovery`,
  `ovs-flow-analysis`, `conntrack-debugging`, `packet-drop-investigation`.

> * Skills MUST only call **named MCP tools** from OKEP-5494 (no free-form

Reviewer: Please remove any references to OKEP-5494. As mentioned earlier, the tools are already implemented and the server can be used. There's no need to mention OKEP-5494 for referring to the tools.

Reviewer: Since there's an option to disable specific tools and/or categories of tools, how will that be factored into the skills?

Author: Since there's an option to disable specific tools and/or categories of tools, how will that be factored into the skills? -- good point, we will clearly have a section in the respective skill.md, that these are the tools required to be enabled/existing, and I believe no one would disable most of the ovn, ovs and minimal kernel tools, 

for example the only thing that I needed to disable is pwru in our ecosystem, because we don't have ebpf support for our servers downstream.

but still lets capture in the skill body that these are the tools required to execute this skill ( if they don't exist don't run, either way it will be explicitly mentioned I think models are smart enough to let the user know)

as a P1, lets capture in docs/ -- a simple mapping of which skill uses which set of tools

Reviewer: > I believe no one would disable most of the ovn, ovs and minimal kernel tools

We can't assume which tools will be disabled and which will be not.

> but still lets capture in the skill body that these are the tools required to execute this skill

Is this captured?

Author: I mean the right place to capture is actuall skill's body, as once skills are in review, the reviewer can suggest, may be adding this tool call would be beneficial or so etc.

Author: we can just mention a statement in OKEP, saying tools required should be mentioned in the skill file -- does this sound acceptable to you

              |  (Cursor / Claude Code / @cursor/sdk)     |
              +--------+---------------------+------------+
                       | loads               | calls
                       v                     v
              +-----------------+   +---------------------+
              |   SKILL.md      |   |   MCP Tools         |
              |   (this OKEP)   |   |   (OKEP-5494)       |
              | frontmatter +   |   | resource-get,       |
              | layered steps + |   | ovn-show, ovn-get,  |
              | decision tree   |   | ovn-trace,          |
              |                 |   | ovs-ofctl-...,      |
              | references --------->  tcpdump, pwru, ... |
              | tools by name   |   |                     |
              +-----------------+   +---------------------+
                       \                  /
                        v                v
                       ovnkube-node pods, NB/SB DBs,
                       OVS, kernel, NIC
```

A skill is a deterministic recipe over the MCP tool catalogue. If the
catalogue changes, the affected skills change in the same PR -
enforced by CI.

> ## Initial skill catalogue thoughts

Reviewer: nit:
```suggestion
## Initial skill catalogue
```

Author: done

| `network-policy-debugging`  | Traffic unexpectedly blocked/allowed | resource-list NetworkPolicy, ovn-get ACL/Address_Set/Port_Group, ovn-trace, ovn-lflow-list, ovs-appctl-ofproto-trace |
| `external-connectivity`     | North-south egress/ingress, SNAT/DNAT issues | ovn-get NAT/Logical_Router/Static_Route, ovn-trace, ovs-vsctl-show, get-ip route, get-iptables nat, get-nft, tcpdump br-ex |
| `node-networking`           | NetworkUnavailable, tunnel down, missing routes | resource-get Node, get-ip address/link/route/rule/neigh, ovs-vsctl-show, ovn-get Chassis, get-iptables, get-nft, pod-logs ovn-controller |
| `ovn-topology-discovery`    | Map / understand the logical network | ovn-show NB/SB, ovn-get LS/LR/LSP/LRP/Chassis/Port_Binding/NAT/Static_Route/Load_Balancer, ovs-vsctl-show |
| `ovs-flow-analysis`         | Logical vs OpenFlow mismatch, stale flows | ovn-lflow-list, ovs-ofctl-dump-flows, ovs-appctl-ofproto-trace, ovn-trace, pod-logs ovn-controller |
| `conntrack-debugging`       | NAT broken, stale entries, table-full, zone conflicts | ovs-appctl-dump-conntrack, get-conntrack -L/-S/-C, ovs-ofctl-dump-flows ct, ovn-lflow-list ct |

> | `packet-drop-investigation` | Silent drops, unknown drop point | tcpdump, ovs-ofctl-dump-flows drop, ovs-appctl-ofproto-trace, ovn-trace, ovs-appctl-dump-conntrack, get-iptables, get-nft, pwru |

Reviewer: Some of the tools have changed. Update them accordingly.

Reviewer: Some of these tools share symptoms like `pod-to-pod-connectivity`, `packet-drop-investigation` and `network-policy-debugging`. It would be good to mention which specific symptom will be used to decide which skill to use. Or may be add another column which mentions when not to use the skill.

Author: > Some of these tools share symptoms like pod-to-pod-connectivity, packet-drop-investigation and network-policy-debugging. It would be good to mention which specific symptom will be used to decide which skill to use. Or may be add another column which mentions when not to use the skill. 

isn't second column already capturing this **When to use**

 the cases when not to use the skill is very hard to tell, and imo we shouldn't be doing this, 
 for example network-policy-debugging, when will you not use it, host level misconfiguration, kernel issues, ovn db corruption, and list might go on. and not required I guess. 
 
 as all we are trying to give the agent which is using the MCP server a deterministic direction + more context, and not limiting agent's capability to use other tools if required.

Reviewer: The above mentioned skills share a similar symptoms. Even one of the examples in the OKEP shows that one can be used instead of the other. So let's either try to make it very clear which one to use when and may be which one not to use when, OR drop the the similar ones in phase 1.

Author: answered here 
https://github.com/ovn-kubernetes/ovn-kubernetes-mcp/pull/52#discussion_r3726946226

* In-tree under `skills/`, reviewed by an OVN-Kubernetes maintainer (or
  the layered-feature owner once feature skills land).
* MCP tool catalogue changes that affect skills must update both in the
  same PR; CI rejects PRs that leave a skill referencing a removed tool.

# Implementation Details

Phase-1 deliverables:

1. Author the initial set of `SKILL.md` files under canonical `skills/`
   (one directory per skill from the catalogue above). Add
   `skills/README.md` as the human-readable index + format reference.
2. Add `.cursor/skills` and `.claude/skills` symlinks.
3. Add a short docs page on pointing Cursor / Claude Code / `@cursor/sdk`
   at the server and the available skills.

In Phase-1, skill correctness is enforced by PR review (well-formedness)
plus the contributor's eval run (behaviour, see Testing Strategy). There
is no automated skill gate yet.

Phase-2 deliverables:

1. **`skills-lint` (`make skills-lint`, wired into CI).** A static checker
   that parses each `SKILL.md`, validates frontmatter, resolves every tool
   reference against the MCP server's tool registry, rejects free-form
   bash and write verbs, and enforces the reporting block and line cap.

>    Implement it in Go by importing the server's `RegisteredToolNames()` so

Reviewer: There's no `RegisteredToolNames()` function currently. Also this references something which might change in the future, either function name or its behavior. Better to rephrase this without referring to any function name explicitly.

Author: Rephrased it, this is phase - 2 though
